src/frequenz/sdk/timeseries/_moving_window.py

```diff
@@ -0,0 +1,170 @@
+# License: MIT
+# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
+
+"""A data window that moves with the latest datapoints of a data stream."""
+
+from __future__ import annotations
+
+import asyncio
+import logging
+from collections.abc import Sequence
+from datetime import datetime, timedelta
+from typing import Union
+
+import numpy as np
+from frequenz.channels import Receiver
+from numpy.typing import ArrayLike
+
+from frequenz.sdk.timeseries import Sample
+from frequenz.sdk.timeseries._ringbuffer import OrderedRingBuffer
+
+log = logging.getLogger(__name__)
+
+
+class MovingWindow(Sequence):
+    """
+    A data window that moves with the latest datapoints of a data stream.
+
+    After initialization the `MovingWindow` can be accessed by an integer
+    index or a timestamp. A sub window can be accessed by using a slice of integers
+    integers or timestamps.
+
+    Note that a numpy ndarray is returned and thus users can use
+    numpys operations directly on a window.
+
+    **Example1** (calculating the mean of a time interval):
+
+    ```
+    window = MovingWindow(size = 100, resampled_data_recv = resampled_data_recv)
+
+    time_start = datetime.now()
+    time_end = time_start + timedelta(minutes=5)
+
+    # ... wait for 5 minutes until the buffer is filled
+    await asyncio.sleep(5)
+
+    # return an numpy array from the window
+    a = window[time_start:time_end]
+    # and use it to for example calculate the mean
+    mean = a.mean()
+    '''
+
+    **Example2** (create a polars data frame from a `MovingWindow`):
+
+    ```
+    import polars as pl
+
+    # create a window that stores two days of data
+    # starting at 1.1.23 with samplerate=1
+    window = MovingWindow(size = (60 * 60 * 24 * 2), sample_receiver)
+
+    # wait for one full day until the buffer is filled
+    asyncio.sleep(60*60*24)
+
+    # create a polars series with one full day of data
+    time_start = datetime(2023, 1, 1)
+    time_end = datetime(2023, 1, 2)
+    s = pl.Series("Jan_1", mv[time_start:time_end])
+    ```
+    """
+
+    def __init__(
+        self,
+        size: int,
+        resampled_data_recv: Receiver[Sample],
+        sampling_period: timedelta,
+        window_alignment: datetime = datetime(1, 1, 1),
+    ) -> None:
+        """
+        Initialize the MovingWindow.
+
+        This method creates the underlying ringbuffer and starts a
+        new task that updates the ringbuffer with new incoming samples.
+        The task stops running only if the channel receiver is closed.
+
+        Args:
+            size: The number of elements that are stored.
+            resampled_data_recv: A receiver that delivers samples with a
+                given sampling period.
+            sampling_period: The sampling period.
+            window_alignment: A datetime object that defines a point in time to which
+                the window is aligned to modulo window size.
+                (default is midnight 01.01.01)
```

Comment:
This is an odd default. Also, it is not entirely clear what this means to me from the docs, maybe it would be good to put some example and clarifying why this default makes sense, as it doesn't look obvious to me.

Reply:
That is inherited from the ring buffer, where we all had understanding trouble :). Will try enhance the docs.

Reply:
Maybe forward to the `RingBuffer` docs if it is already well explained in there.

Reply:
I added an example to the class documentation and referred to that doc there :)